Test forward reference annotations

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#53713 Test forward reference annotations**

Differential Revision: [D26946847](https://our.internmc.facebook.com/intern/diff/D26946847)

Commit:
Test forward reference annotations

[ghstack-poisoned]

jamesr66a commented:
Some of the tests are failing. I think you'll need to gate the tests that rely on `from __future__ import annotations` behind `if sys.version_info >= (3.7)`. Having tests that also exercise this with explicit string annotations would also help coverage as well

Commit:
Update on "Test forward reference annotations"


Differential Revision: [D26946847](https://our.internmc.facebook.com/intern/diff/D26946847)

[ghstack-poisoned]

Commit:
Update on "Test forward reference annotations"


Differential Revision: [D26946847](https://our.internmc.facebook.com/intern/diff/D26946847)

[ghstack-poisoned]

Commit:
Update on "Test forward reference annotations"


Differential Revision: [D26946847](https://our.internmc.facebook.com/intern/diff/D26946847)

[ghstack-poisoned]

walterddr commented on test/fx/test_future.py:
```diff
@@ -0,0 +1,48 @@
+from __future__ import annotations    # type: ignore[attr-defined]
```
this file is not covered by test/run_test.py list. thus it will not run during CI. 

as a matter of fact, none of the tests under test/fx/ folder is included in test/run_test.py. wondering if this is intended?

jamesr66a replied:
Tests in subfolders are generally imported from a "driver" file (test_fx.py) via import statments like these: https://github.com/pytorch/pytorch/blob/bea3cb70693ee85d3f7435efa7232c7fd2fe50cd/test/test_fx.py#L30

It seems that all of those files are imported, less `test_fx_const_fold`. I'll ping the original author

jamesr66a replied:
In fact, for `jit` tests, there's an explicit error condition when trying to run files in the `jit` subfolder: https://github.com/pytorch/pytorch/blob/bea3cb70693ee85d3f7435efa7232c7fd2fe50cd/test/jit/test_logging.py#L11. I might go through and add that to those files for `fx` as well

jamesr66a replied:
Filed an issue: https://github.com/pytorch/pytorch/issues/54071 I'll get around to it eventually

walterddr replied:
sounds good. yeah I realized it was transitively imported via test_fx.py this morning so I created D27080955 to fix the issue. as it turns out the problem is purely internal only
